File: test/hydra/cmds/status_test.go
package cmds

import (
	"os"
	"runtime"
	"strings"
	"testing"
	"time"

	"github.com/micro-plat/hydra"
	"github.com/micro-plat/hydra/hydra/servers/http"
	"github.com/micro-plat/hydra/test/assert"
)

func Test_status_Normal_running(t *testing.T) {
	resetServiceName(t.Name())
	execPrint(t)
	//正常
	defunc, fileCallback := injectStdOutFile()
	defer defunc()

	var app = hydra.NewApp(
		hydra.WithServerTypes(http.API),
		hydra.WithPlatName("xxtest"),
		hydra.WithSystemName("apiserver"),
		hydra.WithClusterName("c"),
	)
	hydra.Conf.API(":19020")

	//1. 删除服务
	os.Args = []string{"xxtest", "remove"}
	app.Start()
	time.Sleep(time.Second * 2)

	//2. 安装服务
	os.Args = []string{"xxtest", "install", "-r", runRegistryAddr, "-c", "c"}
	go app.Start()
	time.Sleep(time.Second * 2)

	//2. 启动服务
	os.Args = []string{"xxtest", "start"}
	go app.Start()

	time.Sleep(time.Second)

	//3. 检查服务状态
	os.Args = []string{"xxtest", "status"}
	go app.Start()

	time.Sleep(time.Second * 2)

	//4. 关闭服务状态
	os.Args = []string{"xxtest", "stop"}
	go app.Start()
	time.Sleep(time.Second * 2)

	//5. 清除服务
	os.Args = []string{"xxtest", "remove"}
	go app.Start()

	time.Sleep(time.Second * 2)
	bytes, err := fileCallback()
	if err != nil {
		t.Error(err)
		return
	}
	line := string(bytes)
	result := strings.Contains(line, "Starting") && strings.Contains(line, "running...")
	assert.Equal(t, true, result, "正常服务运行状态")

	time.Sleep(time.Second)
}

func Test_status_Not_installed(t *testing.T) {

	resetServiceName(t.Name())
	execPrint(t)

	//未安装服务
	defunc, fileCallback := injectStdOutFile()
	defer defunc()

	var app = hydra.NewApp(
		hydra.WithServerTypes(http.API),
		hydra.WithPlatName("xxtest"),
		hydra.WithSystemName("apiserver"),
		hydra.WithClusterName("c"),
	)
	hydra.Conf.API(":19010")

	//1. 清除服务(保证没有服务安装)
	os.Args = []string{"xxtest", "remove"}
	go app.Start()
	time.Sleep(time.Second * 2)

	//2. 启动服务
	os.Args = []string{"xxtest", "status"}
	go app.Start()
	time.Sleep(time.Second * 2)

	bytes, err := fileCallback()
	if err != nil {
		t.Error(err)
		return
	}
	line := string(bytes)

	if runtime.GOOS == "linux" || runtime.GOOS == "darwin" {
		//unbuntu/centos
		result := strings.Contains(line, "sudo") || strings.Contains(line, "Service is not installed")
		assert.Equal(t, true, result, "未安装的服务状态")
	}
	if runtime.GOOS == "windows" {
		result := strings.Contains(line, "not exist as an installed service")
		assert.Equal(t, true, result, "未安装的服务状态")
	}

	time.Sleep(time.Second)
}

File: test/hydra/cmds/status_test.py
# -*- coding: utf-8 -*-  

import os 
import test_all
from test_help import test,execute,runApp
from test_help import ZKAddress

testName = os.path.basename(__file__)

@test(testName)
def test_status_Normal_running():
    try:
        #0.停止并清楚遗留内容
        args = ["stop"]
        runApp(args)
        args = ["remove"]
        runApp(args)
        
        #1.安装服务
        args = ["install","-r",ZKAddress,"-c","c"]
        response = runApp(args)
        #print("install",response)
        if not "OK" in response:
            return u"安装服务失败"

        #2.启动
        args = ["start"]
        response = runApp(args)
        #print("start",response)
        if not "OK" in response:
            return u"启动服务失败"

        #3.停止
        args = ["status"]
        response = runApp(args)
        #print("status",response)
        if not ("Running" in response):
            return u"status服务失败"

        #3.停止
        args = ["stop"]
        response = runApp(args)
        #print("stop",response)
        if not "OK" in response:
            return u"停止服务失败"

    except Exception as err :
        return "error:"+err.message
        
    finally:    
        #4.删除
        args = ["remove"]
        response = runApp(args)
        #print("remove",response)

        if not "OK" in response:
            return u"删除服务失败"



@test(testName)
def test_status_Normal_notrunning():
    try:
        #0.停止并清楚遗留内容
        args = ["stop"]
        runApp(args)
        args = ["remove"]
        runApp(args)

        #1.安装服务
        args = ["install","-r",ZKAddress,"-c","c"]
        response = runApp(args)
        #print("install",response)
        if not "OK" in response:
            return u"安装服务失败"

        args = ["stop"]
        runApp(args)

        #2.status
        args = ["status"]
        response = runApp(args)
        #print("status",response)
        if not "Stopped" in response:
            return u"status服务失败"
 
    except Exception as err :
        return "error:"+err.message
        
    finally:    
        #4.删除
        args = ["remove"]
        response = runApp(args)
        #print("remove",response)

        if not "OK" in response:
            return u"删除服务失败"



@test(testName)
def test_status_not_installed():
    #1. 清理服务，避免其他遗留存在服务
    runApp(["remove"])

    #2.启动服务
    args = ["start"]
    response = runApp(args)
    
    if not ("not installed" in response):
        return u"未安装服务验证"


def main():
     execute(testName)
 
if __name__ == "__main__":
    main()